### ISSUE TITLE
DISCUSS: How to handle breaking changes

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#12004]
There are quite some disagreement about Minetest development when it comes to breaking changes.

I define a "breaking change" as any intentional (i.e. not buggy) change to Minetest that will in some way, shape or form cause breakage to existing mods or games from a previous version. A breakage can also happen when a function was renamed or its return value was changed in an incompatible manner.
Breakage includes obvious things like crashes but if we’re being pedantic it would also include noteworthy graphical changes that modders from the past version could not foresee. One example for this would be the object crosshair.

Now the question is, how exactly should Minetest development deal with breaking changes in future?

My aim of this issue is that we end up with some (semi-?)official document, something like the "roadmap" on which there is broad consensus. 

----

I want to kick off the discussion with several basic ideas (which all need fleshing out):

1) Forbid all breaking changes except for major releases (i.e. 6.0.0, 7.0.0, etc.)
2) Forbid only major breaking changes for non-major releases (i.e. 5.5.0, 5.6.1, etc.)
3) Drop all pretense, embrace the breakage and always make major version releases for even the smallest breakage (SemVer logic)
4) Decide everything on a case-to-case basis
5) (add your own idea here)

As a game maker and modder, I tend to option 1. Although I have the energy to keep mods compatible with latest Minetest, many others do not have that patience, and for good reasons. And at one point, you just want a game to be *finished* and never touch again. Imagine a game that is completely finished and tested and 100% bugfree. A breaking change on the Minetest side essentially a rugpull, introucing bugs to a game with no fault of your own. The downside of option 1 for core development is that some features might be postponed for too long. However, I think compability can still be provided in a lot of cases if we just try hard enough.

As for option 2, the downside is, like said before, the rugpull problem. But it's obviously easier for coredevs. It's bad for modders because basically any minor release could break anything of your game. On the other hand, Minetest releases are rare in general, but I'm not sure if that is a strong excuse. Which leads me to …

Option 3. Under option 3, both versions 5.3.0 and 5.4.0 would have been major releases 6.0.0 and 7.0.0 instead because they broke things. The list of incompabilities that came with version 5.4.0 is surprisingly long. Only versions 5.1.0 and 5.2.0 would have been "proper" minor releases (if I read the Changelog correctly).
So we would be at version 7.0.0 right now. The benefit of option 3 is that this is purely a change on how versions are *named*; the development itself doesn't have to change at all. Also, it makes it much clearer for modders and game authors to declare compability and to prepare for the future. It's easier to say "My game is compatible with version 5.X.X" than "My game is compatible with version 5.4.0 and all later versions in the 5.X.X branch, EXCEPT when there will be an unexpected API breakage in a minor release in the future that is impossible to predict, in which case it is not compatible". :P
Note that accepting option 3 would very probably mean the next version must be 6.0.0. Also, remember, under SemVer, "major version" does NOT mean "big release with many changes" but only "something is incompatible now".

I think option 4 is what we have right now atm, which is the worst. It seems there is no clear consistent rule. This is not really helpful IMO because then we will still have endless discussions like before.

----

Here are some suggestions that I think are no-brainers / we already do but I think should be mentioned explicitly anyway for completeness' sake:

* Patch releases must NEVER come with intentional API breakage
* Every API breakage, no matter how small, must be documented prominently in the Changelog

----

These are my thoughts on this, please add your thoughts as well.

**Comment by sfan5** [comment]
see also my comment here https://github.com/minetest/minetest/issues/11989#issuecomment-1024900496

> I define a "breaking change" as any intentional (i.e. not buggy) change to Minetest that will in some way, shape or form cause breakage to existing mods or games from a previous version

This definition is extremely useless. Take a diff of the `builtin` folder between any two minor version. If you see a variable, function or table which was removed, renamed or it's layout changed no matter how internal or undocumented it is then congrats, that's a breaking change!
It is of course not reasonable to consider that a breaking change, but your definition does so.

You also make no mention of deprecation periods which is pretty much the consensus on how to change things *without* a major release.

> The downside of option 1 for core development is that some features might be postponed for too long.

This downside is way too big to be mentioned along the way inside a paragraph.
If development is slowed down <b>any</b> further by means or more regulations there is a very real possibility that Minetest as a project will just die. You'll have to factor this into your proposal.

**Comment by erlehmann** [comment]
@sfan5 you seem to insist on that breaking changes *have* to happen as frequently as they do now – even though I outlined a way in  how to avoid them. I am quoting myself to make this discussion more readable:

> @sfan5 for many API changes it would be way less disruptive if they were made opt-in by creating a new function with a numeric suffix that increases for every incompatible API change.
> 
> Having functions with names like `minetest.get_player_control2` or `minetest.dynamic_add_media3` may look ugly, but:
> 
> * It would enable a clean separation of old and new functions, helping deprecation.
> * Documentation for an existing function would never go out of date that way.
> * Developers could easily search any mod or even ContentDB for usage.

What exactly keeps Minetest from doing it like that?

**Comment by sfan5** [comment]
Everything:
* We don't want to (aka „consensus“)
* It's more effort
* Incompatible changes are done for a reason, in almost all cases it's not _possible_ to to make the change opt-in (short of just not adding a new feature, if you dislike features I recommend staying on an old version)

**Comment by corarona** [comment]
I was going to post this in the other issue but then I saw this and it probably fits better here. (Hence the quotes from there ( https://github.com/minetest/minetest/issues/11989 )

> Please open another issue for that; users should have an option to decide whether they want their maps aggressively upgraded.

No thanks – I see how issue "discussions" go here. We'll just have to hope they don't break more things i guess. Or maybe it's best to just stay on 5.4. now idk we'll see.

 But I don't think talking here will change anything. The general idea is clearly "we do what we want and if things break it's everyone else's responsibility to fix their stuff". 

>    Yet, any discussion of non-vanilla clients in the forum is forbidden and they are rarely, if ever, asked for advice.

That is indeed one reason I don't really post there but it's not just hacked client discussions: one rule in the forums is literally  "Warnings and (temporary or permanent) bans may be given for failing to follow the rules, or for *any other reason*". It really doesn't get any more arbitrary than that. (Do I get banned for acknowledging hacked clients exist now? who knows lol). It's a bit like that joke "the problem with censorship is XXXXXXXX".

Doesn't exactly make you feel like anyone even wants you to talk. 

To be honest all official minetest things feel super restrictive and authoritarian to me with the general hostile atmosphere that goes with it.

> If it is official policy to not care about compability, then Minetest should at least be honest about it and literally make a major release every release, i.e. 6.0.0, 7.0.0, 8.0.0, etc., in accordance to SemVer, unless there really is no compability break for once. :P

Yeah. That's what I meant with "it should be communicated clearly". On the other hand for mod devs and admins it's probably best to not update under these conditions  unless there are important reasons to do so (like some of the *really* bad bugs in 5.3).

I think it boils down to the fact that the impact on anyone who is not a core dev seems to be a minor concern at best. And yes that's what this is about and not who is "technically" right. [ Hint: this is what happens when you only talk to core devs because everyone else got silenced ]. I realize it isn't *that* extreme but I know I'm not the only one who feels that way.

You can go ahead, delete this and ban me to oblivion if the queen doesn't like my nose :P.

I'm back to read only anyways.

**Comment by erlehmann** [comment]
@sfan5 thanks for the quick answer, but only one of the three arguments makes sense to me:

> * We don't want to (aka „consensus“)

That is a reason why it *is* not done right now – but not a reason why it *can* not be done.

It literally boils down to “Did someone suggest something? LALALA WE CAN NOT HEAR YOU HAHA!”.

Opinions can change in the face of new evidence, like “the current approach is unworkable for game & mod devs”.

> * It's more effort

That is trivially true. It is always more effort. Which is why having a few more coredevs would help here IMO.

What if I offered a patch to keep the old behaviour of `minetest.get_player_control()` and introduce `minetest.get_player_control2()”? Would you – or any core developer – approve? Somehow I doubt it.

> * Incompatible changes are done for a reason, in almost all cases it's not possible to to make the change opt-in (short of just not adding a new feature, if you dislike features I recommend staying on an old version)

This is why I focused specifically on API functions that get new behaviour or different arguments or return codes.

**Comment by appgurueu** [comment]
>     * Incompatible changes are done for a reason, in almost all cases it's not _possible_ to to make the change opt-in (short of just not adding a new feature, if you dislike features I recommend staying on an old version)

Staying on an older version has multiple problems though:

- Not all bugfixes are backported to older versions;
- If you want newer mods (that often require newer versions for the features) to work with older mods (that are broken by the newer versions) to work together, it gets tricky and you effectively have to update the older mods

**Comment by erisMT** [comment]
@erlehmann @corarona Friendly reminder that all this time spent arguing could have been better used to apply changes in your games/ mods, then going about your day.

The core devs don't think there's anything to address, and I believe they are right.

**Comment by corarona** [comment]
> @erlehmann @corarona Friendly reminder that all this time spent arguing could have been better used to apply changes in your games/ mods, then going about your day.

I don't believe I have argued at all to be honest. And yes I agree it probably was an utter waste of time. This does however make my point in a fairly poetic way :).

**Comment by corarona** [comment]
To be more clear: If people want to stop you discussing things in a issue that's literally named DISCUSS that probably says it all.

**Comment by erisMT** [comment]
> > @erlehmann @corarona Friendly reminder that all this time spent arguing could have been better used to apply changes in your games/ mods, then going about your day.
> 
> I don't believe I have argued at all to be honest.

I meant arguing in general, erlehmann seems to have been at odds with the core devs for a while, not you specifically.
> And yes I agree it probably was an utter waste of time. This does however make my point in a fairly poetic way :).

I disagree, you are just antagonizing them for no reason; do you think they'd have gone to these lengths if they didn't deem you worthy of discussing this?

**Comment by sfan5** [comment]
@corarona Your baseless claims of people being silenced or banned does not contribute to constructive discussion either.
If you want to improve the chances of achieving **anything** with this issue then it is very much in your interest to get back on topic. This goes for anyone.

**Comment by erlehmann** [comment]
> Incompatible changes are done for a reason, in almost all cases it's not possible to to make the change opt-in

@sfan5 I looked at a few recent merged pull requests that changed behaviour compared to 5.4.1:

* https://github.com/minetest/minetest/pull/11924 could have just not changed behaviour at literally no cost.
* https://github.com/minetest/minetest/pull/11950 looks like breaking changes in the name of security.
* https://github.com/minetest/minetest/pull/11872 fixes a signed integer overflow (undefined at c level).
* https://github.com/minetest/minetest/pull/11867 looks like breaking changes in the name of security.
* https://github.com/minetest/minetest/pull/11827 made empty formspecs not display, technically breaking them.
* https://github.com/minetest/minetest/pull/11770 fixes a crash but used wrong boundaries even after I pointed it out.
* https://github.com/minetest/minetest/pull/11766 supports mods that assigned textures to models in erroneous ways.
* https://github.com/minetest/minetest/pull/11550 added a new API with the name of the old API, could have not be done.

So among these 8 things we have:

* 2 breaking changes in the name of security (#11950, #11867)
* 3 changes that turn behaviour that is useless under any circumstance into sensible behaviour (#11872, #11766, #11827)
* 3 changes that lead to breakage and could have *easily* been avoided in ways that I laid out (#11924, #11700, #11550)

From this, it seems to me that about half of the breaking changes are easily avoidable in ways that I have laid out.

Curiously, those breaking changes are in fact the ones that generated the most complaints and subsequent discussion.

The other half seems to be only security-relevant fixes and stuff that is so useless that no one would even want to rely on it.

@sfan5 surely this is only a superficial view of the whole issue – but now I have clearly separated “breaking changes” into three different subsets and laid out how the ones that result in the most problems are the ones that are most easily to avoid. Does that maybe change your opinion?

**Comment by corarona** [comment]
> @corarona Your baseless claims of people being silenced or banned does not contribute to constructive discussion either. If you want to improve the chances of achieving **anything** with this issue then it is very much in your interest to get back on topic. This goes for anyone.

I don't think I have claimed anyone got banned  though. The problem isn't necessarily that people actually get banned but that they have to live in constant fear of it. I can tell you I feel that way and as I said I know other people do too.

As for silencing people: The nature of it that they are silent after it i.e. How would you even notice you (inadvertently) silenced them? 

I'm telling you I feel this place is pretty hostile and I have done my best to tell you why I think that is. Now you obviously chose to tell yourself that it can't be true but you know you can't really decide what I feel can you? And again: I am not the only one who feels that way.

I wasn't going to say any more but I do feel what was said here was a bit of a straw man.

**Comment by Zughy** [comment]
I think core devs are trying their best to keep compatibilities as much as they can, even in cases I don't agree with. I personally find this issue pointless, as I'd like to open MT GitHub to see how Minetest is doing, not to read some drama or the umpteenth unnecessary wall of text that basically nobody can or want to read (Wuzzy, please man try to wrap up your thoughts in general).

I think that in general the approach of people like @erlehmann @appgurueu and @corarona is noxious for the project, because it mentally impacts both core devs and contributors. If retrocompatibility is that important to you, please consider the idea of making a fork, and leave Minetest alone. Because you're everywhere, and picking everything to death only has the effect of making people quit.

Also, for what it's worth, I'm sorry for what you've been going through in these "issues", sfan

**Comment by corarona** [comment]
> I disagree, you are just antagonizing them for no reason; do you think they'd have gone to these lengths if they didn't deem you worthy of discussing this?

I'm not sure what you're saying here. Sorry I troubled you to go through the extraordinary lengths of reading half a page!?

However I can tell you that people do not generally say or do things for no reason. My reasons were to point out a problem that by it's nature is invisible to you.

Now I'm sorry you felt antagonized, and yeah maybe I should have worded some of it nicer, then again if someone said something like that to me I (hope I) would be very concerned if there is some truth to  it because – as I pointed out – this isn't generally something you can easily know.

I just felt like I should at least try and point it out since it has been bothering me for a while. Sadly the reactions are not exactly surprising.

**Comment by TSafa-23** [comment]
I feel that core devs have somewhat of a sentiment to minor releases, as we have stayed on 4.X for so long, and from where it's going, it'll be a long time for 5.X as well.

My point is, instead of arbitrarily deciding when the next major release is, why not just have a pre-defined period and one major change _planned_ for it, and be willing to accept other major changes with that release under the guises that they are quote-on-quote "stable"? Six months, eight months, twelve months are pretty fair intervals for releases that will give modders time to adjust their mods if the new releases happen to break compatibility.

**Comment by erisMT** [comment]
(offtopic)

> > I disagree, you are just antagonizing them for no reason; do you think they'd have gone to these lengths if they didn't deem you worthy of discussing this?
> 
> I'm not sure what you're saying here. Sorry I troubled you to go through the extraordinary lengths of reading half a page!?

I did skim quickly through your post, but I guess your point just struck me as being a bit accusatory towards the team, which seems to just be discussing this issue with you, not threatening bans.

> However I can tell you that people do not generally say or do things for no reason. My reasons were to point out a problem that by it's nature is invisible to you.

> I just felt like I should at least try and point it out since it has been bothering me for a while. Sadly the reactions are not exactly surprising.

I'm sorry it came across as if in my post I invalidated your feelings, I didn't mean it that way; you can of course feel some hostility, and that's alright. Just try to be constructive, as sfan5 said. peace \/

(/offtopic)

**Comment by appgurueu** [comment]
> I think core devs are trying their best to keep compatibilities as much as they can, even in cases I don't agree with.

The core devs definitely have the best intentions at heart, but don't necessarily see things from a modder's or player's perspective. Who would have thought that not returning a string anymore in some obscure function which was never supposed to break anything might cause a bunch of mods to break? Monitoring and reporting such breakage is crucial.

> I personally find this issue pointless, as I'd like to open MT GitHub to see how Minetest is doing, not to read some drama or the umpteenth unnecessary wall of text that basically nobody can or want to read (Wuzzy, please man try to wrap up your thoughts in general).

Oh all these claims... You're focusing solely on what you consider "progress", only caring about the newest version as you're not maintaining any old codebase like MCL. Minetest would be doing bad if it decided to not care about breakage in minor releases.

If you don't want to read it, don't. That doesn't mean nobody "can or wants to read [it]" though.

> I think that in general the approach of people like @erlehmann @appgurueu and @corarona is noxious for the project, because it mentally impacts both core devs and contributors. If retrocompatibility is that important to you, please consider the idea of making a fork. Picking everything to death only has the effect of making people quit.

Breaking changes actively impact existing codebases, particularly the older ones. Retrocompatibility is important not only to us, but to Minetest in general, and has always been an advantage of many open source software over proprietary competitors. Why make a fork for this when Minetest has reliably guaranteed compatibility for the past years? Forks tend to be a massive mess.

Constantly breaking compatibility has the effect of making people quit as well - just on the other end, with modders quitting. As usual, a compromise must be found. I consider myself not overly "picky" (when opening issues at least) FWIW.

> Also, for what it's worth, I'm sorry for what you've been going through in these "issues", sfan

What has sfan been "going through"? If anything, these issues are at worst a waste of everyone's time, at best helpful in preventing (sometimes) unnecessary breakage.

**Comment by rubenwardy** [comment]
>  one rule in the forums is literally "Warnings and (temporary or permanent) bans may be given for failing to follow the rules, or for any other reason". It really doesn't get any more arbitrary than that. (Do I get banned for acknowledging hacked clients exist now? who knows lol). It's a bit like that joke "the problem with censorship is XXXXXXXX".

The intent of this rule is to allow edge cases to still be moderatored, but given the over rules it's probably not necessarily.
The no cheat rule also feels outdated and [isn't even enforced](https://forum.minetest.net/viewtopic.php?p=396825)

I'll propose some rule changes

**Comment by lhofhansl** [comment]
Not sure it makes sense to comment here after the ... passionate ... discussion.

We could certainly introduce semantic versioning and define what exactly we can break in a patch, minor, and major release.

It would require version/branch management. Are we having a 5.4.6 *and* a 5.5.2 version for those who want bug fixes but not take the breaking changes? 

Then again, I do not think we release often enough to warrant that. 
That could be fixed by introducing train model. The train comes by (say) once per month, every change done, ready, and releasable by then gets on board, all other changes wait for the next train. Obviously only non-breaking changes would get on the regular train.

Anecdote: I once volunteered to managed a large open source project. Before I started we had no clear compatibility guidelines, had a release every 12-18 months. Before each release there was a mad scramble to get stuff (since the next release is so far away) with the inevitable bugs introduced in such large release, followed by many stabilization releases afterwards.
I replaced it with said train model, and we started following SemVer, and everyone was happier, less stressed, quality went up. But it was a lot of work in branch/version management.

Morale of the story. It SemVer and clean guidelines really are better, but it is a lot of work, and we're all part-time volunteers here.
In the end progress is better than stand-still, and I feel Minetest development is already slow. Let's not slow it down further. Please. But let's also not go out of our way to break things.

**Comment by Wuzzy2** [comment]
Note I clarified that I do allow for some esoteric edge cases to not count as breakage. I don't want to imply that every small change is immediately a "breakage". I am also NOT saying that incompatible changes are always forbidden. That's what major releases are for. Of coure, ancient legacy garbage has to be nuked from time to time. I'm not against that. The question is: When and how?

And yeah, I definitely want to hear more opinions about my SemVer idea (option 3) because this option wouldn't even require a change of process, nor would it add bureaucracy, it's just a different "version naming convention". For the future releases, this would probably lead to a lot of major releases at first, if nothing else changes. But should then one particular release actually have absolutely zero breakage, then you could proudly present that as a minor release.

I agree the downside for option 1 is indeed bad. True, adding more regulations would probably get Minetest to a screeching halt. That would indeed be even worse.

**Comment by erlehmann** [comment]
> Of coure, ancient legacy garbage has to be nuked from time to time. I'm not against that. The question is: When and how?

Btw, people often focus on the *ancient* here – but I think the important part is that one should throw out only *garbage*.

“If it ain't broke, don't fix it.” is a tried and tested way of keeping APIs stable over literal decades (see HTTP or XMPP).

> And yeah, I definitely want to hear more opinions about my SemVer idea (option 3) because this option wouldn't even require a change of process, nor would it add bureaucracy, it's just a different "version naming convention". 

From a game & mod development and user perspective think adopting semantic versioning would help a lot.

A reason that has not been mentioned is that Linux distributions that package Minetest may assume semantic versioning.

> I agree the downside for option 1 is indeed bad. True, adding more regulations would probably get Minetest to a screeching halt. That would indeed be even worse.

Do you have any idea why at least for the Lua API, no one actually interacted with my proposal of just not messing with existing functions, but adding numbered ones except for @wsor4035 (who often just downvotes my comments without any explanation)?

**Comment by v-rob** [comment]
Hang on a minute: Let's say we have function `minetest.bla_do_thing(x, y, z)`.  Someone calls it with:

```
local my_table = {1, 2, 3, 4}
minetest.bla_do_thing(unpack(my_table))
```

Now we add a new parameter, `minetest.bla_do_thing(x, y, z, t)`.  Suddenly, their code breaks and we get blamed for it.  Even the most simple changes can result in incompatibilities!  The API is frozen in time forever unless we update the major version.

OK, so we add `minetest.bla_do_thing2(x, y, z, t)`.  Importantly, we have _twice_ the amount of C++ code, one per function (`l_bla_do_thing` and `l_do_bla_thing2`), so _twice_ the likelihood that one gets out of date during, say, a refactoring.  That's bad.  But uh oh, someone has a sandboxed Lua environment (similar to Mesecons luacontroller), but instead of doing a whitelist, they blacklisted every Minetest function they didn't want.  They should have `minetest.bla_do_thing2` sandboxed, but don't since they made their mod for version 28.2 of Minetest, not Minetest 28.3.  It breaks their mod, so we'll have to do Minetest 29.0!

How unreasonable does it get before we call it quits and just increment the major version every release and lose the support of all our modders because development is frozen?  The Window 95 approach to compatibility is unsustainable and unreasonable.

**Comment by v-rob** [comment]
On a separate note:  We do occasionally break mods in minor releases, it's true.  Do we break the majority of mods, or even a large portion?  No.

I used to be a normal user before I was a developer, and not all that long ago either.  Let's say it was 2016, around 0.4.14 or so: I'm browsing through the Mod Releases section.  I find a mod from 2013 for Minetest 0.4.5 or something.  I download it, run it, and it causes an error.  I'm a little put out because the mod looked cool, and I'm no programmer so I can't fix it.  Was I ever surprised?  Not in the least.  The mod's old by this point; why should I expect it to work?  It's probably not very broken, after all, and a few touches could fix it.  This was invariably my reaction.  Looking though this thread, I would have considered it an extreme overreaction, and hence I do now as well.

**Comment by freshreplicant** [comment]
> 

As a player I echo this. Some level of breakage is totally normal and reasonable players will recognise that, especially in a modding ecosystem as diverse and long standing as Minetest. It's already a million times more accessible as a user than modded playthroughs of popular proprietary games where juggling breakages, incompatibility, mod managers and a jumble of random scripts is normal. I honestly think that this isn't as big an issue on the player side as it's made out to be, though I understand that for developers breakages add to the maintenance burden.

**Comment by nikolaus-albinger** [comment]
I agree with v-rob and freshreplicant, there will always be breaks, but most of them are documented so they can be fixed. And most of the time it's not even that much work.

**Comment by LoneWolfHT** [comment]
I was/am completely fine with older mods not working. I am also fine with it in the case of games, but I feel like we can help with that at least a little.

What if you added support for Minetest running games with older Minetest versions?
AFAIK you just need the older exe, libraries, and builtin for a decent chance of it working/not breaking your client. You could also implement an automatic updater while you're at it.

**Comment by erlehmann** [comment]
> The mod's old by this point; why should I expect it to work?

Other games manage to not break existing mods. OpenTTD, for example, has improved its mod format (NewGRF) a lot compared to older versions – and as far as I am aware, if some old mod turns out not to work, they will usually investigate and fix it in a newer release.

By the way, a web site made 25 years ago still works, too – despite the web having gained a lot of new features in the meantime.

> I don't mean to say that breakage is a good thing, but it's inevitable at some point or another. We should try to break as little as possible without entering the pedantic and unreasonable.

I consider this a rhetorical trick: People who either do not want or can not implement compatible solutions *always* say it is “reasonable” or “pragmatic”, even when it is not. But @sfan5 stated it multiple times in the past: Compatibility means effort.

> I honestly think that this isn't as big an issue on the player side as it's made out to be, though I understand that for developers breakages add to the maintenance burden.

That is the core issue: The maintenance effort does not vanish. Only in one case it is done once in the engine, while in other cases every single game and mod needs to be fixed. I think it is *reasonable* and *pragmatic* to fix an issue once and for all in a central place instead of requiring a lot of people to do duplicated busywork. Obviously, the people who would actually have to put in the work (the developers of the Minetest engine) often disagree, if pushing the issue downstream looks like it would not cause too much friction. This also means that game and mod code that actually updates properly accumulates workarounds.

**Comment by v-rob** [comment]
> > The mod's old by this point; why should I expect it to work?
> 
> Other games manage to not break existing mods. OpenTTD, for example, has improved its mod format (NewGRF) a lot compared to older versions – and as far as I am aware, if some old mod turns out not to work, they will usually investigate and fix it in a newer release.

They have different developers, mods, requirements, and mindsets.  We do not have the resources (i.e. developer time) to fix some compatibility problems, and for problems like `player:get_player_control()`, we do not think that we should generally support mods relying on very obviously buggy behaviour, because that corrupts our modding API and makes life worse for us and other people in the long run.

> By the way, a web site made 25 years ago still works, too – despite the web having gained a lot of new features in the meantime.

False: websites that used Microsoft JScript or VBScript or ActiveX or bugs in Internet Explorer or used things like Adobe Flash or Java applets all do not work anymore.  They relied on non-standard extensions or bugs.  If mods break because they relied on `player:get_player_control()` returning `""` for non-players, or relying on an undocumented function or parameter, that is in no way, shape, or form Minetest Engine's fault, so why are we being blamed?

> > I don't mean to say that breakage is a good thing, but it's inevitable at some point or another. We should try to break as little as possible without entering the pedantic and unreasonable.
> 
> I consider this a rhetorical trick: People who either do not want or can not implement compatible solutions _always_ say it is “reasonable” or “pragmatic”, even when it is not. @sfan5 has stated it multiple times in the past: _Compatibility means effort._

It's reasonable to a different definition.  Your definition is that it's not reasonable because it breaks mods.  Our definition is that is it reasonable because the mods where already broken by relying on broken behaviour.  If a C++ program relies on undefined behaviour, technically, the C++ program is entirely undefined.  It's not up to the compiler to support this program relying on the undefined behaviour.  Therefore, I consider it quite "reasonable" according to this definition.

> Obviously, the people who would actually have to put in the work (the developers of the Minetest engine) often disagree, if pushing the issue downstream looks like it would not cause too much friction. This also means that game and mod code that actually updates properly accumulates workarounds.

"Workarounds"?  By supporting mods relying on buggy behaviour, _Minetest Engine_ has to add workarounds.  By fixing the buggy behaviour, we are _removing_ workarounds for everybody.

**Comment by Zughy** [comment]
quoting https://github.com/minetest/minetest/issues/12004#issuecomment-1024907073

> If development is slowed down any further by means of more regulations there is a very real possibility that Minetest as a project will just die.

plus what other core devs said, basically on the same page. Closing

**Comment by appgurueu** [comment]
Great, that way the entire discussion will just be resumed whenever another breaking change is PR'd.

**Comment by Wuzzy2** [comment]
QA is very important, still. We should have at least SOME standards. Okay, I kind of agree now it is probably not helping if we just forbid ANY breakage outright. However, that doesn't mean it's a good idea to not even try to avoid breakage.

It would be terrible if we tell PR authors they don't have to care about breakage at all. That's a great way to frustrate modders and game authors.

I don't like it if games randomly change their behavior or even break just because the player updated to a new minor (!) Minetest version. It's OK and expected for major Minetest versions, but for minor versions, too? What's the point of minor versions then?

While I agree that a 100% strict requirement on breakage might cause problems sometimes, I think it is still a good idea to at least require from PR authors to at least TRY to avoid breakage, as long it is within a reasonable degree. Like, when a breaking change can be averted by simple fallback / backwards-compability code, then just require that from the PR author, so that the breaking change becomes a non-breaking change. Then everyone wins.

The object crosshair is the perfect example of a (technically) breaking change (existing games suddenly reverted to the default crosshair when pointing entities, with no fault of their own) that could have been averted by fallback behavior (i.e. when `object_crosshair.png` is missing).

So basically, all I am asking for now is this:

When a PR introduces a breaking change, no matter how small, first check if the breakage can be *reasonably* averted. If yes, ask the PR author to add fallback / backwards-compability code. "reasonably" here means when you don't expect the fallback to be absurdly complicated, requires changing tons of files and functions, etc.

Does this sound more OK to you?
